### PR TITLE
kmdb(1) for aarch64

### PR DESCRIPTION
**Pull request by richlowe**
This is the halting, in-situ, version of the kernel debugger.

A description of the current implementation is in:
   usr/src/uts/aarch64/kdi/README.md

It is not at all perfect, but I think is at least often useful at this point.

I am not sure about the behaviour when debugging the debugger with the debugger, and can still induce a lot of spew about `mdb_ks` in some situations (but I remember these on intel too).

Unfortunately `::noducttape` which tries to test this gets optimized away, and I can't think of anything else that would trap in the right way which does not.

### CODE REVIEW

**Comment by richlowe** [comment]
@hadfl @r1mikey @citrus-it how does this go on rpi?  I think it might at least make a valiant effort, but I expect problems

**Comment by r1mikey** [comment]
Wow, this is huge!

I'm starting to look through it and will test on Raspberry Pi 4 in a few hours.

File: usr/src/cmd/mdb/aarch64/mdb/mdb_aarch64util.c [hunk 1]
```diff
@@ -51,3 +160,249 @@ mdb_aarch64_next(mdb_tgt_t *t, uintptr_t *p, kreg_t pc, mdb_instr_t curinstr)
...
+}
+
+int
+mdb_aarch64_kvm_framev(void *arglim, uintptr_t pc, uint_t argc,
+    const long *argv, const mdb_tgt_gregset_t *gregs)
+{
+	/*
+	 * Historically adb limited stack trace argument display to a fixed-
+	 * size number of arguments since no symbolic debugging info existed.
+	 * On amd64 we can detect the true number of saved arguments so only
```
Comment (r1mikey) [inline usr/src/cmd/mdb/aarch64/mdb/mdb_aarch64util.c]: On aarch64.

File: usr/src/cmd/mdb/common/kmdb/kctl/kctl_main.c [hunk 2]
```diff
@@ -521,8 +521,6 @@ kctl_modload_activate(size_t memsz, const char *cfg, uint_t flags)
 		return (rc);
 	}
 
-	kctl.kctl_flags = flags;
```
Comment (r1mikey) [inline usr/src/cmd/mdb/common/kmdb/kctl/kctl_main.c]: Is this removal intentional?

File: usr/src/cmd/mdb/common/mdb/mdb_fmt.c [hunk 3]
```diff
@@ -798,13 +798,22 @@ mdb_fmt_print(mdb_tgt_t *t, mdb_tgt_as_t as,
 				rvalue = u.i4;
 				break;
 			case 8:
+				/*
+				 * XXXKDI: cannot even refer to floating point
```
Comment (r1mikey) [inline usr/src/cmd/mdb/common/mdb/mdb_fmt.c:802]: Do we need fp context save and restore? Floating point is always present in some form on armv8+.

File: usr/src/cmd/mdb/intel/amd64/kmdb/kaif_invoke.S [hunk 4]
```diff
@@ -53,7 +53,7 @@ copyargs:
 
 	/*
 	 * This is going to be fun.  We were called with the function pointer
-	 * in in %rsi, argc in %rdx, and a pointer to an array of uintptr_t's
+	 * in %rsi, argc in %rdx, and a pointer to an array of uintptr_t's
```
Comment (r1mikey) [inline usr/src/cmd/mdb/intel/amd64/kmdb/kaif_invoke.S:56]: Upstream first?

File: usr/src/cmd/mdb/intel/kmdb/kaif.c [hunk 5]
```diff
@@ -26,9 +26,9 @@
  */
 
 /*
- * The debugger/"PROM" interface layer
+ * The debugger/"PROM debugger" interface layer
```
Comment (r1mikey) [inline usr/src/cmd/mdb/intel/kmdb/kaif.c:29]: Also upstream first?

File: usr/src/cmd/mdb/intel/mdb/kvm_amd64dep.c [hunk 6]
```diff
@@ -104,57 +104,57 @@ kt_stackv(uintptr_t addr, uint_t flags, int argc, const mdb_arg_t *argv)
...
-	(int (*)())(uintptr_t) mdb_tgt_null,	/* t_add_iowapt */
-	(int (*)())(uintptr_t) mdb_tgt_null,	/* t_add_sysenter */
-	(int (*)())(uintptr_t) mdb_tgt_null,	/* t_add_sysexit */
-	(int (*)())(uintptr_t) mdb_tgt_null,	/* t_add_signal */
-	(int (*)())(uintptr_t) mdb_tgt_null,	/* t_add_fault */
-	kt_getareg,				/* t_getareg */
-	kt_putareg,				/* t_putareg */
-	mdb_amd64_kvm_stack_iter,		/* t_stack_iter */
-	(int (*)())(uintptr_t) mdb_tgt_notsup	/* t_auxv */
+	.t_setflags = kt_setflags,
```
Comment (r1mikey) [inline usr/src/cmd/mdb/intel/mdb/kvm_amd64dep.c]: Also upstream first?

File: usr/src/cmd/mdb/intel/mdb/kvm_ia32dep.c [hunk 7]
```diff
@@ -104,57 +104,57 @@ kt_stackv(uintptr_t addr, uint_t flags, int argc, const mdb_arg_t *argv)
...
-	(int (*)())(uintptr_t) mdb_tgt_null,	/* t_add_iowapt */
-	(int (*)())(uintptr_t) mdb_tgt_null,	/* t_add_sysenter */
-	(int (*)())(uintptr_t) mdb_tgt_null,	/* t_add_sysexit */
-	(int (*)())(uintptr_t) mdb_tgt_null,	/* t_add_signal */
-	(int (*)())(uintptr_t) mdb_tgt_null,	/* t_add_fault */
-	kt_getareg,				/* t_getareg */
-	kt_putareg,				/* t_putareg */
-	mdb_ia32_kvm_stack_iter,		/* t_stack_iter */
-	(int (*)()) mdb_tgt_notsup		/* t_auxv */
+	.t_setflags = kt_setflags,
```
Comment (r1mikey) [inline usr/src/cmd/mdb/intel/mdb/kvm_ia32dep.c]: Also upstream first?

File: usr/src/uts/aarch64/kdi/README.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+We swap the entire vector rather than patching a single one, or adjusting
+indirection through function pointers since it's proved the most foolproof in
+practice.
+
+## Entering the debugger
+
+We always enter kmdb via an exception.  For manual entry we enter via a `svc`
+from EL1 back to EL1, similar to a system call.  Other entry occurs has we
+handle the relevant ARM synchronous exceptions.
```
Comment (r1mikey) [inline usr/src/uts/aarch64/kdi/README.md]: Should that be "Other entry occurs as"?

File: usr/src/uts/aarch64/kdi/README.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+exceptions.
+
+We then set `MDSCR.SS` and call `kmdb_dpi_resume_master` which returns back
+out to the kernel, immediately takes a single step trap, and comes right back
+again, where we undo our register changes.
+
+# Debug exceptions and their masking v. single step
+
+There are various ways in which debug exceptions may be masked (for the
+purposes of this, I'm including in "masked" the case where the CPUU never
```
Comment (r1mikey) [inline usr/src/uts/aarch64/kdi/README.md]: :%s/CPUU/CPU/

File: usr/src/uts/aarch64/sys/privregs.h [hunk 10]
```diff
@@ -129,14 +142,44 @@ struct regs {
 	stp	x17, x18, [sp, #REGOFF_PC];	\
 	stp	fp, x17, [sp, #REGOFF_SAVFP];
 
+#define	__SAVE_REGS_EL1				\
```
Comment (r1mikey) [inline usr/src/uts/aarch64/sys/privregs.h:145]: This could make running softints at do_spl-time easier - will pick that up again one this lands.

File: usr/src/uts/intel/kdi/kdi_idt.c [hunk 11]
```diff
@@ -365,8 +365,10 @@ kdi_deactivate(void)
  * can prove to us that the kernel is prepared to handle the trap, we'll assume
  * there's a problem and will give the user a chance to debug it.
  *
- * If we return 2, then the calling code should restore the trap-time %cr3: that
- * is, it really is a kernel-originated trap.
+ * If we return false, handle this trap ourselves.
```
Comment (r1mikey) [inline usr/src/uts/intel/kdi/kdi_idt.c:368]: Also upstream first?

File: usr/src/uts/intel/asm/thread.h [hunk 12]
```diff
@@ -46,8 +46,8 @@ struct _kthread;
  * Yuck.
  */
 
-extern __GNU_INLINE struct _kthread
-*threadp(void)
+extern __GNU_INLINE struct _kthread *
```
Comment (r1mikey) [inline usr/src/uts/intel/asm/thread.h:49]: Also upstream first?

File: usr/src/uts/armv8/virt/io/virt.c [hunk 13]
```diff
@@ -74,8 +74,8 @@ _ischar()
 	return (!(UARTFR & UARTFR_RXFE));
 }
 
-static void _reset(bool poff) __NORETURN;
-static void _reset(bool poff)
+void __NORETURN
```
Comment (r1mikey) [inline usr/src/uts/armv8/virt/io/virt.c:77]: These machine-specific resets are something I want to rework in the not-too-distant future. We have a hard dependency on PSCI (which is not a problem), so there's no real need to hide this away as machine-specific.

Comment (r1mikey) [reply usr/src/uts/armv8/virt/io/virt.c:77]: For the record, that's just an idle comment, not a review comment.

File: usr/src/cmd/mdb/aarch64/kmdb/kvm_isadep.c [hunk 14]
```diff
@@ -0,0 +1,500 @@
...
+	    (psr & PSR_A ? 'A' : 'a'),
+	    (psr & PSR_I ? 'I' : 'i'),
+	    (psr & PSR_F ? 'F' : 'f'),
+	    aarch64_el_name(psr & PSR_M_MASK));
+
+	mdb_printf("%%esr    = 0x%0?p\n", gregs->kregs[KREG_ESR]);
+	mdb_printf("%%trapno = %0?p\n", gregs->kregs[KREG_TRAPNO]);
+}
+
+#if 0
```
Comment (r1mikey) [inline usr/src/cmd/mdb/aarch64/kmdb/kvm_isadep.c]: Yeah, no ioports on Arm :)

File: usr/src/cmd/mdb/aarch64/kmdb/kaif.c [hunk 15]
```diff
@@ -0,0 +1,663 @@
...
+	kmdb_dpi_resume_master(); /* Run the next instruction and come back */
+
+	/* restore the single step setting and mask settings */
+	kmdb_dpi_set_register("spsr", spsr);
+	write_mdscr_el1(read_mdscr_el1() & ~MDSCR_SS);
+
+	return (0);
+}
+
+/*ARGSUSED*/
```
Comment (r1mikey) [inline usr/src/cmd/mdb/aarch64/kmdb/kaif.c]: Delete the comment? All arguments are used.

**Review by r1mikey — APPROVED** [review]
Looks good, though I lack debugger experience.

I've dropped a few comments inline - most are just comments or optional cleanup. There are a few files modified here that might be better to be done upstream so that a later merge is less painful, but that's entirely up to you.

I'll spin this up on Pi4 a little later.

File: usr/src/cmd/mdb/common/kmdb/kctl/kctl_main.c (hunk 2 above)
Comment (richlowe) [reply usr/src/cmd/mdb/common/kmdb/kctl/kctl_main.c]: No it wasn't, I put this back.  Oops

File: usr/src/cmd/mdb/common/mdb/mdb_fmt.c (hunk 3 above)
Comment (richlowe) [reply usr/src/cmd/mdb/common/mdb/mdb_fmt.c:802]: The arm64 gcc and ARM architecture are very literal about not using floating point (like in the kernel and kmdb).  If we say "Don't use floating point", you can't even mention a floating point type.  It is the equivalent of the FPU being turned off.

File: usr/src/cmd/mdb/intel/kmdb/kaif.c (hunk 5 above)
Comment (richlowe) [reply usr/src/cmd/mdb/intel/kmdb/kaif.c:29]: I was hoping to do this as part of trying to clear up these layering descriptions, but ran into a wall because they don't make sense, and I've been having to figure out what they were probably meant to be based on how they acted on SPARC.

I left the minor cleanups behind on the basis that they improve things, and don't distract.

I can split from this commit, but I'm not sure I can upstream preductively.

File: usr/src/cmd/mdb/intel/mdb/kvm_amd64dep.c (hunk 6 above)
Comment (richlowe) [reply usr/src/cmd/mdb/intel/mdb/kvm_amd64dep.c]: This is illumos "16315 mdb ops vectors should use c99 initialization"

File: usr/src/cmd/mdb/aarch64/kmdb/kvm_isadep.c (hunk 14 above)
Comment (richlowe) [reply usr/src/cmd/mdb/aarch64/kmdb/kvm_isadep.c]: Sigh.  did I miss deleting some of that?  I'll get that

File: usr/src/cmd/mdb/aarch64/kmdb/kaif.c (hunk 15 above)
Comment (richlowe) [reply usr/src/cmd/mdb/aarch64/kmdb/kaif.c]: I'll get this, and clean up the others `ARGSUSED` in files I touched to use `__unused` unless they're really ignoring _every_ argument

**Comment by r1mikey** [comment]
Testing on Pi4 seems equivalent to Qemu for the lightweight poking around I've done.

I used booted into the early debugger (`-kd`) in Qemu, on Pi4 is just used `mdb -K`.

Leaving the debugger via `CTRL+D` or `::quit` isn't working (in both Qemu and Pi4).

```
[3]>
kmdb: Do you really want to reboot? (y/n) y
Rebooting...

Welcome to kmdb
kmdb: cannot modify read-only variable 'mdb_ks'
[3]>
```

Getting back into the OS via `:c` works on both machines.

Other than that, using `mdb -K` to get into the debugger, `:c` to leave it, basic lightweight operations seem to be working the same as they do on Qemu (pgrep, what's, findstack, findleaks, breaking in syscall entry, single stepping).

This is going to be incredibly useful.

**Review by r1mikey — APPROVED** [review]
I'm sure there will be further refinements as things mature, but I tested basic equivalence between Qemu and Raspberry Pi4, and noticed no machine-specific differences.

This is going to be incredibly useful as we work to mature the port.

**Comment by richlowe** [comment]
> ```
> [3]>
> kmdb: Do you really want to reboot? (y/n) y
> Rebooting...
> 
> Welcome to kmdb
> kmdb: cannot modify read-only variable 'mdb_ks'
> [3]>
> ```

The 2nd warning here is because we're re-loading an already loaded kmdb, and attempting to load mdb_ks on itself can't work.

I have fixed the "not even trying to reboot" case (I forgot to look for x86 v. SPARC weirdness that needed ifdef changes).  You may panic rebooting if you reboot before PSCI exists, the same as with `panic`. (that is, booting `-kd` and immediately doing `$q` will panic, getting further and doing `$q` will reset properly).

**Comment by richlowe** [comment]
Re: upstream stuff. I have upstreamed the c99 init, and will definitely be not merging this until I have merged it with that, so that should make things much more contained.

I'm currently hunting for where serial breaks on the console are fully handled.  Gosh STREAMS can be a mess.

**Comment by richlowe** [comment]
That last push enables entering the debugger non-fatally when receiving an abort sequence.

Usually this would be a serial break on the console, I can't work out how to get that into qemu via stdio, so I switched to the alternate break `~^B` with `abort_enable/W 2` to test this.

**Comment by r1mikey** [comment]
Last two diffs tested, including a reboot from a non-zero CPU.

Just a note that PSCI has no constraint on the CPU that a reboot operation can be invoked from, so you could simplify that codepath on aarch64 if you wanted.

**Comment by r1mikey** [comment]
> That last push enables entering the debugger non-fatally when receiving an abort sequence.
> 
> Usually this would be a serial break on the console, I can't work out how to get that into qemu via stdio, so I switched to the alternate break `~^B` with `abort_enable/W 2` to test this.

I've tested the break functionality successfully.

Reboot is a bit odd (both Qemu and Pi4) - if we're on CPU 0 things work fine. If we're on any other CPU, no reboot happens. I think this is a problem dancing over to other CPUs in that codepath. Since PSCI can call reboot from any CPU we should probably just call the `_reset` in the aarch64 case.

**Comment by richlowe** [comment]
I merged this onto arm64-gate (merged onto illumos-gate), and force-pushed this.
If nobody objects, I'll push this soon.
